File: solutions/go/list-ops/1/list_ops.go
package listops

// IntList is an abstraction of a list of integers which we can define methods on
type IntList []int

func (s IntList) Foldl(fn func(int, int) int, initial int) int {
	for i := 0; i < len(s); i++ {
    	initial = fn(initial, s[i])
    }
    return initial
}

func (s IntList) Foldr(fn func(int, int) int, initial int) int {
	for i := len(s) - 1; i >= 0; i-- {
		initial = fn(s[i], initial) 
	}
	return initial 
}

func (s IntList) Filter(fn func(int) bool) IntList {
	res := make(IntList, 0)
    for _, x := range(s) {
        if fn(x) {
            res = append(res, x)
        }
    }
    return res
}

func (s IntList) Length() int {
	return len(s)
}

func (s IntList) Map(fn func(int) int) IntList {
	res := make(IntList, 0)
    for _, x := range(s) {
        res = append(res, fn(x))
    }
    return res
}

func (s IntList) Reverse() IntList {
	for i := 0; i < len(s) / 2; i++ {
        s[i], s[len(s) - i - 1] = s[len(s) - i - 1], s[i]
    }
    return s
}

func (s IntList) Append(lst IntList) IntList {
	return append(s, lst...)
}

func (s IntList) Concat(lists []IntList) IntList {
	for _, x := range(lists) {
        s = append(s, x...)
    }
    return s
}

File: solutions/python/list-ops/1/list_ops.py
def append(list1, list2):
    result = list1.copy()
    for x in list2:
        result.append(x)
    return result


def concat(lists):
    result = []
    for item in lists:
        if isinstance(item, list):
            for sub_item in item:
                result.append(sub_item)
        elif item != None:
            result.append(item)
    return result


def filter(function, lst):
    result = []
    for x in lst:
        if function(x):
            result.append(x)
    return result


def length(lst):
    return len(lst)


def map(function, lst):
    result = []
    for x in lst:
        result.append(function(x))
    return result


def foldl(function, lst, initial):
    for x in lst:
        initial = function(initial, x)
    return initial


def foldr(function, lst, initial):
    for x in lst[::-1]:
        initial = function(initial, x)
    return initial



def reverse(lst):
    return lst[::-1]
